vision_evaluation/evaluators.py

```diff
@@ -1046,3 +1046,64 @@ def add_predictions(self, predictions, targets):
             gt_mask = np.asarray(gt_mask)
             mean_l1 = np.abs(pred_mask.astype(np.float)-gt_mask.astype(np.float)).mean()
             self._metric_sum += mean_l1
+
+
+
+class GroupWiseEvaluator(Evaluator):
+    """
+    Group Wise Evaluator
+    """
+
+    def __init__(self, evaluator):
+        super().__init__()
+        self.evaluator = evaluator
+
+    def _group_by(self):
+        """
+            Group predictions and targets by target_task and group labels.
+            Example: IC on target_task = 'ethnicity' and group = 'gender'
+
+            store dict will save target labels and predictions corresponding to each group label.
+
+            store = {(target_task, group) : {
+                "targets":[]
+                "predictions":[]
+            }}
+        """
+
+        self.store = collections.defaultdict(
+            lambda: collections.defaultdict(list))
+
+        for (target_task, group), prediction in zip(self.targets, self.predictions):
+            self.store[(target_task, group)]['predictions'].append(prediction)
+            self.store[(target_task, group)]['targets'].append(target_task)
+
+    def add_predictions(self, predictions, targets):
+        """ Evaluate a batch of predictions.
+        Args:
+            predictions: the model output numpy array. Shape (N, num_class)
+            targets: the golden truths. Shape (N,2). The first column is the target_task and the second column is the group label.
+        """
+
+        assert len(predictions) == len(targets)
+        self.targets = targets
+        self.predictions = predictions
+
+    def get_report(self):
+
+        # Step 1: Group targets and predictions by target_task and group
+        self._group_by()
+
+        # Step 2: For each combination of target_task and group labels, compute the metric as per the given evaluator.
+        self.metrics = collections.defaultdict()
```

Comment:
no need to make it an instance variable. just 'metrics' instead of 'self.metrics' is ok.

Reply:
updated